src/doc/style-guide/src/expressions.md

```diff
@@ -536,6 +540,70 @@ if a_long_expression
 {
     ...
 }
+
+if let Some(a) = b
+    && another_long_expression
+    || a_third_long_expression
+{
+    ...
+}
+
+if let Some(relatively_long_thing)
+    = a_long_expression
+    && another_long_expression
+    || a_third_long_expression
+{
+    ...
+}
+
+if some_expr
+    && another_long_expression
+    && let Some(relatively_long_thing)
+        = a_long_expression
+    || a_third_long_expression
+{
+    ...
+}
+```
+
+A let-chain control line is allowed to be formatted on a single line provided
+it only consists of two clauses, separated by `&&`, with the first being an
```

Comment:
We'd explicitly documented the desire to limit this to `&&`, so I captured it as such here. I actually don't know whether any other binops are supported within chains, but I think it would be in our best interest to not scope the binops here unless we feel strongly that `&&` vs `||` should make a difference. I included an example with `||` below to demonstrate that the concerning scenario I raised against style option 2 (which led us to the eventual option 5 captured in this PR) could still exist if we restrict to `&&`

Reply:
>  I actually don't know whether any other binops are supported within chains

`||` isn't allowed in a chain if there are any `let`s. Just FYI, the examples shown here aren't valid due to that rule.

Reply:
Thanks that's helpful. I think one consideration though is what constitutes valid in the style/formatting context. I.e. at what stage does this get rejected, will it cause the parser to error out prior to creating the AST or does it occur later? Is there a potential for the supported operators to change in the future?

There's existing precedent within the Style Guide that covers "invalid" flavors of syntax, e.g.

https://github.com/rust-lang/rust/blob/master/src/doc/style-guide/src/items.md#where-clauses
```
// Note that where clauses on `type` aliases are not enforced and should not
// be used.
type Foo<T>
where
    T: Bound
= Bar<T>;
```

Which I always assume was the case because rustfmt has to account for it so long as the AST can still be produced. If `||` is rejected during lexing/parsing and we don't think that'll ever change then I'm happy to remove it, but otherwise my inclination would be to update the examples to remove the `||` while also relaxing the rule to remove the explicit "separated by `&&`"

Reply:
The validation is not done during parsing. It is in the AST validator (which IIRC, is done just after expansion).

The RFC mentioned that `||` might be possible in the future, but I'm not aware of any serious consideration for it currently, and I suspect the addition would require another RFC.